impl/src/main/java/impl/UserAnswerResourceImpl.java

```diff
@@ -43,27 +52,58 @@ public UserAnswerResourceImpl( AnswerDao answerDao) {
-            .switchIfEmpty(error(new WebException(HttpResponseStatus.NOT_FOUND, ANSWER_NOT_FOUND)))
+            .switchIfEmpty(exception(() -> new WebException(HttpResponseStatus.NOT_FOUND, ANSWER_NOT_FOUND)))
             .flatMap(storedAnswer -> {
-            if(auth.getUserId() != storedAnswer.getUserId()) {
-                return error(new WebException(HttpResponseStatus.FORBIDDEN, NOT_OWNER_OF_ANSWER));
-            }
-            return answerDao.updateAnswer(auth.getUserId(),  answerId, answer)
-                .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_UPDATE_ANSWER, throwable)));
-        });
+                if (auth.getUserId() != storedAnswer.getUserId()) {
+                    return error(new WebException(HttpResponseStatus.FORBIDDEN, NOT_OWNER_OF_ANSWER));
+                }
+                return answerDao.updateAnswer(auth.getUserId(), answerId, answer)
+                    .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_UPDATE_ANSWER, throwable)));
+            });
     }
 
     @Override
     public Observable<Void> deleteAnswer(Auth auth, long answerId) {
         return answerDao.getAnswerById(answerId)
             .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_GET_ANSWER_FROM_DATABASE, throwable)))
-            .switchIfEmpty(error(new WebException(HttpResponseStatus.NOT_FOUND, ANSWER_NOT_FOUND)))
+            .switchIfEmpty(exception(() -> new WebException(HttpResponseStatus.NOT_FOUND, ANSWER_NOT_FOUND)))
             .flatMap(storedAnswer -> {
-            if(auth.getUserId() != storedAnswer.getUserId()) {
-                return error(new WebException(HttpResponseStatus.FORBIDDEN, NOT_OWNER_OF_ANSWER));
-            }
-            return answerDao.deleteAnswer(auth.getUserId(), answerId)
-                .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_DELETE_ANSWER, throwable)));
-        });
+                if (auth.getUserId() != storedAnswer.getUserId()) {
+                    return error(new WebException(HttpResponseStatus.FORBIDDEN, NOT_OWNER_OF_ANSWER));
+                }
+                return answerDao.deleteAnswer(auth.getUserId(), answerId)
+                    .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_DELETE_ANSWER, throwable)));
+            });
+    }
+
+    @Override
+    public Observable<Void> upVoteAnswer(Auth auth, long answerId) {
+        return handleVote(new Vote(auth.getUserId(), answerId, 1));
+    }
+
+    @Override
+    public Observable<Void> downVoteAnswer(Auth auth, long answerId) {
+        return handleVote(new Vote(auth.getUserId(), answerId, -1));
+    }
+
+    private Observable<Void> handleVote(Vote newVote) {
+
+        return first(answerDao.getAnswerById(newVote.getAnswerId()).flatMap(answer -> {
```

Comment:
i think that using first to wrap around a so large observable can make it hard to read. Either move out the observable given to ´first´in a variable or do as the implementation in first does

Reply:
I'll break out some things and see what you think.

Reply:
So, I removed first/then and broke some stuff into methods. We'll see if it improves readability.